packages/@react-aria/dialog/src/useDialog.ts

```diff
@@ -41,7 +41,7 @@ export function useDialog(props: AriaDialogProps, ref: RefObject<HTMLElement>):
       // or announce that it has opened until it has rendered. A workaround
       // is to wait for half a second, then blur and re-focus the dialog.
       let timeout = setTimeout(() => {
-        if (document.activeElement === ref.current) {
+        if (document.activeElement === ref.current || !ref.current.contains(document.activeElement)) {
```

Comment:
This fixes the issue and seems reasonable, but I had hoped that something like FocusScope's `useFocusContainment` would've handled keeping the focus within the Dialog on mount, especially if it is the currently active scope. I guess a Dialog has an implicit `contains focus` behavior on it...

Reply:
I think I may have been able to correctly diagnose the problem.

1. When the Modal opens, the new Dialog sets focus to itself, which triggers `onFocus` within the `useFocusContainment` hook of the Popover Dialog containing the "Open modal" Button that is closing. 
2. At line 278 in `onFocus` within the `useFocusContainment` hook, focus gets immediately restored to the "Open modal" button within the Popover Dialog that is closing. https://github.com/adobe/react-spectrum/blob/4386323091e4b9372706895b06d93b271b71975a/packages/%40react-aria/focus/src/FocusScope.tsx#L274-L284
3. When the Popover Dialog closes, focus gets restored to the "Open popover" underneath the Modal Dialog that had received focus in the first place.

The problem seems to be, that `contain` should be set `false` on the FocusScope that is closing, so that the `onFocus` event handler within `useFocusContainment` for the scope that is closing gets removed. There doesn't seem to be an easy way to do this, because a Dialog has no way of knowing whether it is open or closing, that is all handled by the overlay.

Reply:
New approach using `FocusScope.autoFocus` within Dialog: 1bb3bf2e625bb3d259fc0d17126e692efbef44f8